### DISCUSSION TITLE
How to upload image to digital ocean space in nextjs13 new api route ?

### DISCUSSION THREAD

**Discussion by zwelhtetyan** [vercel/next.js#50499] [App Router]
??

**Comment by icyJoseph** [comment]
If you want to do it from the server, create a form so that when the client submits the form, you send a `FormData` object as body to the backend. From there you have to work with the file you receive.

Do you have any SDK for Digital Ocean? I think people often use the aws-sdk right?

- https://docs.digitalocean.com/products/spaces/reference/s3-sdk-examples/ 

## Edit

This works with Node 18.13+:

```js
// app/upload/route.ts
import { S3, PutObjectCommand } from "@aws-sdk/client-s3";
import { File } from "buffer";
import { NextResponse } from "next/server";

const s3Client = new S3({
  forcePathStyle: false,
  endpoint: process.env.SPACES_ENDPOINT!,
  region: "us-east-1",
  credentials: {
    accessKeyId: process.env.SPACES_KEY!,
    secretAccessKey: process.env.SPACES_SECRET!,
  },
});

export async function POST(req: Request) {
  try {
    const form = await req.formData();
    const file = form.get("file-input");

    if (!file) return NextResponse.json({ message: "failure" });

    const isFile = file instanceof File;

    if (!isFile) return NextResponse.json({ message: "failure" });

    const buffer = await file.arrayBuffer();

    const data = await s3Client.send(
      new PutObjectCommand({
        Bucket: process.env.SPACES_NAME,
        Key: "a-name-for-this-element",
        Body: Buffer.from(buffer),
      })
    );

    console.log(data);
    return NextResponse.json({ message: "success" });
  } catch (reason) {
    console.log(reason);
    return NextResponse.json({ message: "failure" });
  }
}

```

You just need to get the API keys and such. I found this one to be a bit easy to miss:

```
SPACES_ENDPOINT=https://{hosting-region}.digitaloceanspaces.com
```

The SPACES_NAME is whatever you called your space.

And the two other keys come from here: https://docs.digitalocean.com/products/spaces/how-to/manage-access/#access-keys

Last but not least, some client code. Server actions would this even easier though.

```jsx
"use client";

export default function Home() {
  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        fetch("/upload", {
          method: "POST",
          body: new FormData(e.currentTarget),
        });
      }}
    >
      <input type="file" name="file-input" />
      <button type="submit">submit</button>
    </form>
  );
}
```

**Reply by icyJoseph** [reply]
@zwelhtetyan updated with actual working code 😅  managed to upload to spaces.

You'll see some warnings regarding `aws-crt`, I think you can ignore those.

**Reply by grinvaldsjanis** [reply]
Please some really working example of serverless upload to presigned url to DO Spaces from next.js app on localhost:3000
Maybe You are lucky - nothing works for me.
